oeps/oep-0011.rst

```diff
@@ -0,0 +1,299 @@
+   coding style across multiple codebases, edX has published an ESLint
+   configuration that provides an enforceable specification. The
+   `edX ESLint Config`_ is made available as an npm package that can be
+   installed into any Open edX package.
+
+5. **Front-end interactivity should be added using Backbone**
+
+   Rationale: There are a large number of JavaScript libraries for
+   building rich client applications, but many of them are too
+   rigid for a flexible, extensible codebase like Open edX. Backbone
+   has the major benefit that it can be added incrementally to existing
+   pages, no matter the framework that is used. It also provides a clean
+   programming model that enforces a clean separation between models
+   and views.
+
+   Note: Backbone is a somewhat old technology in the JavaScript world
+   and has seen a rapid drop off in usage within the community. See
+   :ref:`JavaScript Frameworks` below for the open question of which
+   of the many alternate frameworks should become the new Open edX standard.
+   For now, Backbone is the library of choice because so much of the
+   Open edX codebase uses it.
+
+6. **JavaScript dependencies should be declared via RequireJS**
+
+   Rationale: Open edX is a large, complicated code base, and it is
+   getting more tangled over time. RequireJS allows code to declare
+   the dependencies it requires, rather than forcing every page to load
+   all possible libraries. It also supports intelligent bundling using
+   RequireJS Optimizer, which can package together all required libraries
+   into a single file for faster performance.
+
+7. **CSS should be generated using SASS**
+
+   Rationale: Sass is an extension of CSS that adds power and elegance to
+   the basic language. It makes the maintenance of large CSS files more
+   manageable though the ability to use variables, mixins, imports and more.
+   In particular, it makes theming possible with the ability to override
+   variables that define colors, fonts etc.
+
+   You can find out more about Sass with the official guide
+   `Sass: CSS with Superpowers`_.
+
+Open Questions
+==============
+
+.. _JavaScript frameworks:
+
+1. **Which more modern JavaScript framework should be supported?**
+
+   The JavaScript space is very fast moving as can be seen in
+   `The State of JavaScript 2016 Survey`_. There is a lot of momentum
+   around `React`_, and it appears to have many qualities that would
+   make it a good fit for edX. It is too early to make this call, but
+   plans are being made to evaluate its usage. See `React at edX`_ for
+   more details.
+
+2. **How should offline support be provided?**
+
+   For many mobile users, it would be very useful if they could continue
+   their Open edX courses while offline. This is true both for true mobile
+   apps as well as for responsive web applications. This is an area that
+   has not been explored much up to now, but which will only become more
+   important going forward.
```

Comment:
The last time I did a serious investigation into this (2-3 years ago), it was a minefield of missing functionality and inconsistency across browsers...and getting worse over time rather than better.  It's probably worth looking at again with the advent of service workers and such, but I recommend going into it with healthy skepticism about the odds of being able to create a good offline experience via an actual web browser (vs. a native app which includes HTML rendering capability).